File: server/app/schemas/progress.py
from pydantic import BaseModel
from typing import Optional


class WeeklyPoint(BaseModel):
    week: str
    avg_score: float
    attempts: int


class TaskMetric(BaseModel):
    task_id: str
    task_name: str
    overall_accuracy: float
    total_attempts: int
    current_level: Optional[str]
    pass_rate: float
    last_attempt_result: Optional[str]


class ProgressResponse(BaseModel):
    total_attempts: int
    avg_final_score: float
    pass_rate: float
    weekly_trend: list[WeeklyPoint]
    task_metrics: list[TaskMetric]
    dominant_emotion: Optional[str]

File: client/types/progress.ts
export interface WeeklyPoint {
  week: string;
  avg_score: number;
  attempts: number;
}

export interface TaskMetric {
  task_id: string;
  task_name: string;
  overall_accuracy: number;
  avg_phoneme_accuracy: number | null;
  phoneme_accuracy_count: number;
  total_attempts: number;
  current_level: string | null;
  pass_rate: number;
  last_attempt_result: string | null;
}

export interface Progress {
  total_attempts: number;
  avg_final_score: number;
  avg_phoneme_accuracy: number | null;
  phoneme_accuracy_count: number;
  pass_rate: number;
  weekly_trend: WeeklyPoint[];
  task_metrics: TaskMetric[];
  dominant_emotion: string | null;
}
